### PR TITLE
add whisper-update

### PR DESCRIPTION
**Pull request by AnderEnder**

### CODE REVIEW

File: whisper/src/bin/whisper-update.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+extern crate whisper;
+
+use failure::Error;
+use std::path::PathBuf;
+use std::time::{SystemTime, UNIX_EPOCH};
+use structopt::StructOpt;
+use whisper::point::Point;
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "whisper-dump")]
```
Comment (loony-bean) [inline whisper/src/bin/whisper-update.rs]: typo: dump

Comment (AnderEnder) [reply whisper/src/bin/whisper-update.rs]: copy-paste

File: whisper/src/point.rs [hunk 2]
```diff
@@ -30,3 +32,21 @@ impl Point {
...
+
+impl FromStr for Point {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Point, Error> {
+        let segments: Vec<&str> = s.split(":").collect();
+
+        let (interval, value) = match segments.len() {
+            2 => (segments[0], segments[1]),
+            _ => return Err(format_err!("Can not parse point from string: {}", s)),
```
Comment (loony-bean) [inline whisper/src/point.rs]: typo: Can not

File: whisper/src/point.rs [hunk 3]
```diff
@@ -1,5 +1,7 @@
 use std::io;
 use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use std::str::FromStr;
+use failure::Error;
 
 #[derive(Debug, Clone, Copy)]
 pub struct Point {
```
Comment (andy128k) [inline whisper/src/point.rs]: This structure is already implemented.It's called 'Retention'

Comment (AnderEnder) [reply whisper/src/point.rs]: this is not a retention. This is a Point

File: whisper/src/bin/whisper-update.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    #[structopt(name = "timestamp:value")]
+    points: Vec<Point>,
+}
+
+fn main() -> Result<(), Error> {
+    let args = Args::from_args();
+    println!("whisper-update {}", env!("CARGO_PKG_VERSION"));
+    println!("{:?}", args);
+
+    let now = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() as u32;
```
Comment (loony-bean) [inline whisper/src/bin/whisper-update.rs]:
```
if len(datapoints) == 1:
   timestamp, value = datapoints[0]
   whisper.update(path, value, timestamp)
 else:
   whisper.update_many(path, datapoints)
```

this original (python) logic is missing

File: whisper/src/point.rs [hunk 5]
```diff
@@ -1,5 +1,7 @@
 use std::io;
 use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use std::str::FromStr;
+use failure::Error;
```
Comment (andy128k) [inline whisper/src/point.rs]: failure::Error is good for CLIs but library should define its own errors.

File: whisper/src/bin/whisper-update.rs (hunk 4 above)
Comment (AnderEnder) [reply whisper/src/bin/whisper-update.rs]: It was already discussed.  It should not copy the original logic 1x1.

Comment (loony-bean) [reply whisper/src/bin/whisper-update.rs]: Any reason for breaking compatibility with existing (python) cli tools?
I thought the idea is to provide drop-in replacements for these

Comment (loony-bean) [reply whisper/src/bin/whisper-update.rs]: I'm not agains the new behaviour, but we might choose some new bin name not to confuse the users

Comment (andy128k) [reply whisper/src/bin/whisper-update.rs]: The only issue I see here is possible performance penalty.

Comment (AnderEnder) [reply whisper/src/bin/whisper-update.rs]: > I'm not agains the new behaviour, but we might choose some new bin name not to confuse the users

Behaviour is the same. The difference in custom performance optimisations, which is not very valuable for cli implementation. Also Rust compiler can level down this custom optimisation.

File: whisper/src/point.rs [hunk 6]
```diff
@@ -34,19 +31,19 @@ impl Point {
...
-            _ => return Err(format_err!("Cannot parse point from string: {}", s)),
+            _ => return Err(ParseError::ParsePointError(s.to_string())),
         };
 
-        Ok(Point{
-            interval: interval.parse()?,
-            value: value.parse()?,
+        Ok(Point {
+            interval: interval.parse().map_err(|e| ParseError::ParseIntError(e))?,
+            value: value.parse().map_err(|e| ParseError::ParseFloatError(e))?,
```
Comment (andy128k) [inline whisper/src/point.rs]: .map_err(ParseError::ParseFloatError)

**Comment by AnderEnder** [comment]
reverted cargo_fmt

**Pull request merged**